### PR TITLE
Implemented blocking of UNIQUE constraint/index on nullable column consistently

### PR DESCRIPTION
**Pull request by rohit01010**
### Description

Blocking of UNIQUE constraint/index on nullable column was not implemented consistently. This PR will cover all the different ways to add unique constraint/index on nullable column and will block those statements by giving error. This blocking of such statements will only happen when babelfishpg_tsql.escape_hatch_unique_constraint is not set to ignore.

Updated some test files in which statements trying to create unique index/constraint on nullable columns, as the changes contained in this PR will raise an error for such statements.

Signed-off-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)

### Issues Resolved
BABEL-2619

### Test Scenarios Covered ###
* **Use case based -** YES

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by robverschoor — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 1]
```diff
@@ -876,8 +877,8 @@ pltsql_post_parse_analyze(ParseState *pstate, Query *query, JumbleState *jstate)
 								{
 									ereport(ERROR,
 											(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-											 errmsg("Nullable UNIQUE constraint is not supported. Please use babelfishpg_tsql.escape_hatch_unique_constraint to ignore "
-													"or add a NOT NULL constraint")));
+											errmsg("Nullable UNIQUE constraint is not supported. Please use babelfishpg_tsql.escape_hatch_unique_constraint to ignore "
+												"or add a NOT NULL constraint")));
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Unrelated changes

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 2]
```diff
@@ -893,15 +894,45 @@ pltsql_post_parse_analyze(ParseState *pstate, Query *query, JumbleState *jstate)
 									def->constraints = lappend(def->constraints, get_rowversion_default_constraint(def->typeName));
 								}
 								break;
-							case T_Constraint:
+							case T_Constraint: 
+							{
+								Constraint *c = (Constraint *) element;
+								ListCell   *lc;
+								
+								c->conname = construct_unique_index_name(c->conname, stmt->relation->relname);
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Why do you need to set conname before doing the error checks?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 3]
```diff
@@ -893,15 +894,45 @@ pltsql_post_parse_analyze(ParseState *pstate, Query *query, JumbleState *jstate)
...
+								
+								if (escape_hatch_unique_constraint != EH_IGNORE &&
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: What's the purpose of this code? Please add some comments?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 4]
```diff
@@ -940,7 +971,7 @@ pltsql_post_parse_analyze(ParseState *pstate, Query *query, JumbleState *jstate)
 						/* Check for identity attribute */
 						if (attr->attidentity)
 							seen_identity = true;
-
+
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Unnecessary change

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 5]
```diff
@@ -982,100 +1021,100 @@ pltsql_post_parse_analyze(ParseState *pstate, Query *query, JumbleState *jstate)
...
+								{
+									ereport(ERROR,
+											(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+											errmsg("Nullable UNIQUE constraint is not supported. Please use babelfishpg_tsql.escape_hatch_unique_constraint to ignore "
+												"or add a NOT NULL constraint")));
 								}
+
+								if (rowversion_column_name)
+									validate_rowversion_table_constraint(c, rowversion_column_name);
+							}
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Please fix the formatting for rest of the changes. It's not clear what exactly you've modified in this code apart from the formatting.

**Review by kuntalghosh — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 1 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: resolved

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 5 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Done

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 4 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Removed

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 6]
```diff
@@ -1519,11 +1558,34 @@ has_unique_nullable_constraint(ColumnDef *column)
...
+			default:
+				break;
+		}
+	}
+
+	return !is_notnull;
+}
+
 static bool
 is_nullable_constraint(Constraint *cst, Oid rel_oid)
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c:1593]: You've completely changed the definition of the function. Is that expected? Also, could you please add a top-level comment what the function is trying to do?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 7]
```diff
@@ -1589,7 +1651,7 @@ static bool
 is_nullable_index(IndexStmt *stmt)
```
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/src/pl_handler.c:1663]: Same for this function - changed the definition of the function. Is that expected? Also, could you please add a top-level comment what the function is trying to do?

**Review by kuntalghosh — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 6 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c:1593]: Yes it was expected, as is_nullable_constraint was initially returning true only if all the columns in the constraint is null else false. But expected behaviour should be return true if atleast one of the column is null else false(if all the columns are not null)

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 7 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c:1663]: Yes it was expected, as is_nullable_index was initially returning true only if all the columns in the index is null else false. But expected behaviour should be return true if atleast one of the column is null else false(if all the columns are not null)

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 6 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c:1593]: Added a top-level comment

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 7 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c:1663]: Added a top level comment

**Review by kuntalghosh — APPROVED** [review]
